### PR TITLE
Conda reverse proxy

### PR DESCRIPTION
**Pull request by AaronJackson**
Working towards conda-forge but we can change channels quite easily (I think)

### CODE REVIEW

File: reverse_proxy/defaults/main.yml [hunk 1]
```diff
@@ -5,3 +5,7 @@ zone: tre.example.org
 # cran mirror
 cran_mirror: https://cran.ma.imperial.ac.uk/
 cran_regex: ^/(bin/windows|src)/contrib/(4\.1|4\.3|3\.6)*/*[a-zA-Z0-9_\-\.]*(\.tar\.gz|\.zip|PACKAGES)$
+
+# conda-forge
+conda_repo_regex: ^/conda-forge/(linux-64|linux-aarch64|win-64|noarch)/(current_)*repodata.json$
```
Comment (manics) [inline reverse_proxy/defaults/main.yml:10]: `(current_)*`
Should this be `?` instead of `*`?

File: reverse_proxy/defaults/main.yml [hunk 2]
```diff
@@ -5,3 +5,7 @@ zone: tre.example.org
...
+conda_pkg_regex: ^/conda-forge/(linux-64|linux-aarch64|win-64|noarch)/[a-zA-Z0-9\-\._!]*(\.tar\.bz2|\.conda)$
```
Comment (manics) [inline reverse_proxy/defaults/main.yml:11]: What's the significance of `!` in the regex? Is it used in filenames or does it mean something else?

**Review by manics — COMMENTED** [review]
I'm not enough of a regex expert to understand it all 😃 

`diff -u cran.conf.j2 conda.conf.j2`:
```diff
--- cran.conf.j2	2023-01-09 15:41:48.000000000 +0000
+++ conda.conf.j2	2023-01-09 15:42:00.000000000 +0000
@@ -1,23 +1,23 @@
 server {
   listen 80;
-  server_name cran.{{zone}};
+  server_name conda.{{zone}};
 
   # Internal only endpoint for performing a proxy pass to the cran mirror.
-  location /cran-fetch {
+  location /conda-fetch {
     internal;
-    proxy_pass {{cran_mirror}};
+    proxy_pass https://conda.anaconda.org/;
   }
 
   # Define an EICAR endpoint for verifying that the reverse proxy is
   # scanning correctly.
-  location /cran-fetch/src/contrib/eicar.tar.gz {
+  location /conda-fetch/conda-forge/eicar/download/noarch/eicar.conda {
     internal;
     rewrite_by_lua_block {
       ngx.say("X5O!P%@AP[4\\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*")
     }
   }
 
-  location ~ {{cran_regex}} {
+  location ~ ({{conda_pkg_regex}}|{{conda_repo_regex}}) {
     # ensure we are only processing GET requests
     rewrite_by_lua_block {
       function reject ()
@@ -43,7 +43,14 @@
 
       -- Perform a fetch using the internal cran-fetch endpoint. Everything after the
       -- "/cran-fetch" is passed to the configured cran mirror.
-      local res = ngx.location.capture("/cran-fetch" .. ngx.var.uri);
+      ngx.log(ngx.STDERR, ngx.var.uri)
+      local res = ngx.location.capture("/conda-fetch" .. ngx.var.uri);
+
+      -- Get the headers to ensure we return the correct Content-Type
+      local ctype = res.header['Content-Type']
+      local cenc  = res.header['Content-Encoding']
+      ngx.header["Content-Type"] = ctype
+      ngx.header["Content-Encoding"] = cenc
 
       local path = os.tmpname()
 
@@ -64,7 +71,7 @@
       ngx.header["X-Sophos-Scan-Timer"] = tostring(t_end - t_start) .. " cpu seconds"
 
       if ret == 0 then
-        ngx.say(res.body)
+        ngx.print(res.body)
       else
         ngx.status = ngx.HTTP_BAD_GATEWAY
         ngx.header["Content-Type"] = "text/plain"
```
Do you think it's feasible to have a single parameterised template for this and all future static website proxies?

File: reverse_proxy/defaults/main.yml (hunk 2 above)
Comment (AaronJackson) [reply reverse_proxy/defaults/main.yml:11]: Yes :(

e.g. x264-1!164.3095-h4e544f5_2.tar.bz2

File: reverse_proxy/defaults/main.yml (hunk 1 above)
Comment (AaronJackson) [reply reverse_proxy/defaults/main.yml:10]: Good point!

File: reverse_proxy/molecule/default/verify.yml [hunk 3]
```diff
@@ -150,3 +150,50 @@
...
+  - name: test connection to /conda-forge/noarch/current_repodata.json
+    assert:
+      that: conda_curl_repodata.status == 200
+      fail_msg: "Request to valid repodata was blocked."
+      success_msg: "Request to valid repodata was not allowed."
+
+  - name: perform installation of a package (numpy)
+    shell: |
+      eval "$(/root/miniconda3/bin/conda shell.bash hook)"
+      conda uninstall -y numpy || true
```
[the remark this answers is not in the document]
Comment (AaronJackson) [reply reverse_proxy/molecule/default/verify.yml:188]: nice, I like that idea

File: reverse_proxy/defaults/main.yml (hunk 2 above)
Comment (manics) [reply reverse_proxy/defaults/main.yml:11]: It's to denote an epoch https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#version-ordering

**Comment by AaronJackson** [comment]
@manics I've created a generic template which covers both conda and cran - good suggestion, thanks!

**Review by manics — APPROVED** [review]
LGTM.

A possible future refinement is to not hard code-cran/conda-forge, and instead make the list of proxies completely parameterised, e.g.:
```yaml
proxy_configs:
  conda-forge:
    generic_name: conda
    upstream_endpoint: https://conda.anaconda.org/
    eicar_path: conda-forge/eicar/download/noarch/eicar.conda
    endpoint_regex: "({{ conda_pkg_regex }}|{{ conda_repo_regex }})"
  cran:
    generic_name: cran
    upstream_endpoint: "{{ cran_mirror }}"
    eicar_path: src/contrib/eicar.tar.gz
    endpoint_regex: "{{ cran_regex }}"
```

```yaml
- name: proxy config
  template:
    src: generic.conf.j2
    dest: /etc/nginx/sites-enabled/{{ item.key }}.conf
  loop: "{{ proxy_configs | dict2items }}"
```
and reference `{{ item.value }}` in the template. Or even better, don't have separate files, just have a for loop in a single template (makes it easier to delete a proxy on an existing deployment)

This makes it trivial to add a new proxy such as another conda channel- the role is just a generic proxy template, and the playbook configures the proxies

**Pull request merged**